flyteplugins/go/tasks/pluginmachinery/ioutils/remote_file_output_reader.go

```diff
@@ -3,77 +3,235 @@ package ioutils
 import (
 	"context"
 	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
 
 	"github.com/pkg/errors"
 
 	"github.com/flyteorg/flyte/flyteidl/gen/pb-go/flyteidl/core"
 	"github.com/flyteorg/flyte/flyteplugins/go/tasks/pluginmachinery/io"
+	"github.com/flyteorg/flyte/flyteplugins/go/tasks/pluginmachinery/k8s"
 	"github.com/flyteorg/flyte/flytestdlib/storage"
 )
 
-type RemoteFileOutputReader struct {
-	outPath        io.OutputFilePaths
+type errorRetriever interface {
```

Comment:
`RemoteFileOutputReader` already implements the `OutputReader` interface, see [here](https://github.com/flyteorg/flyte/blob/1e54d21c4d4ee74245f799a57b4bb8a5534e8368/flyteplugins/go/tasks/pluginmachinery/io/iface.go#L32).

I would prefer if this could be refactored in a way that doesn't require introducing another interface for this.

Reply:
`IsError` also seems to serve the same purpose as `HasError`, `ReadError` the same purpose as `GetError`.

Reply:
@fg91  `RemoteFileOutputReader` still only implements `OutputReader` and not the `errorRetriever`.

An output reader has 2 functionalities: output reading and error reading. I refactored `RemoteFileOutputReader` so that it has an error retriever member, which can be either `singleFileErrorRetriever` or `earliestFileErrorRetriever`. Both `singleFileErrorRetriever` and `earliestFileErrorRetriever` implement the `errorRetriever` interface. It is the standard strategy pattern.

It looks like you are thinking of having a `MultiFileRemoteFileOutputReader` that extends from `RemoteFileOutputReader`. I don't quite like that (and also commented on the RFC about this sometime back), as we are creating a new output reader variety for each error reading strategy, instead of creating new error reader varieties.

Here is what I can do to avoid defining a new interface: I'll refactor `OutputReader` interface to inherit from an `ErrorReader` interface, so we won't introduce something new and use `ErrorReader` interface for the `errorRetriever` member (may be call it `errorReader`).

Let me do this and if you are still not comfortable with it, we can perhaps chat about it.

Reply:
See https://github.com/flyteorg/flyte/pull/5795/commits/0a12e95f4768a57f9770b3f36ec7849827acfe5a

Reply:
yeah agree with that, no to MultiFileRemoteFileOutputReader i think.  just the error side is enough.